src/inferencer/constraintStore.ts

```diff
@@ -1,11 +1,90 @@
+import { Type, Variable, Primitive } from '../types'
+
 export const constraintStore = new Map()
 
-export function updateTypeConstraints(newConstraintLhs: number, newConstraintRhs: number | string) {
-  // step 3a. Add new constraint to constraintStore map
-  // e.g. T1 = T2, T2 = number
-  constraintStore.set(newConstraintLhs, newConstraintRhs)
+export function updateTypeConstraints(newConstraintLhs: Type, newConstraintRhs: Type) {
+  // console.log('updateTypeConstraints')
+  // console.log(newConstraintLhs + ' ' + newConstraintRhs)
+  solveConstraint(newConstraintLhs, newConstraintRhs)
+  // if (detectError(constraintStore)) console.log('Error! Rule 9')
+}
+
+function solveConstraint(constraintLhs: Type, constraintRhs: Type) {
+  // check if both key and value are not base types (Rule 1)
+  if (!is_base_type(constraintLhs) && !is_base_type(constraintRhs)) {
+    // check if key is not a type variable and value is a type variable (Rule 2)
+    if (is_base_type(constraintLhs) && is_type_variable(constraintRhs)) {
+      constraintStore.set(constraintRhs, constraintLhs)
+    }
+    // Rule 3
+    else if (
+      is_type_variable(constraintLhs) &&
+      constraintStore.get(constraintRhs) !== undefined &&
+      is_type_variable(constraintStore.get(constraintRhs)) &&
+      constraintStore.get(constraintRhs) === constraintLhs
```

Comment:
I did a quick google, I believe typescript's  `===` compare objs by reference. i.e. 2 separately instantiated objs even w same values (in our case) will return false. If that's the case, I believe this comparator isn't what we want. See: https://stackoverflow.com/questions/43634154/how-typescript-comparing-two-objects

Reply:
Okay I'm not sure if our both objects are separately instantiated, but definitely better to play safe and compare by values instead of reference. E.g. we could do instead `constraintStore.get(constraintRhs).getId() === constraintLhs.getId()`